core/modules/wsched/ChunkTaskList.cc

```diff
@@ -0,0 +1,393 @@
+            if (newActive == _activeChunk) {
+                newActive = _chunkMap.end();
+            }
+            _chunkMap.erase(_activeChunk);
+        }
+
+        _activeChunk = newActive;
+        if (newActive == _chunkMap.end()) {
+            return false;
+        }
+        newActive->second->movePendingToActive();
+        newActive->second->setActive();
+    }
+
+    // Advance through chunks until READY or NO_RESOURCES found, or until entire list scanned.
+    auto iter = _activeChunk;
+    ChunkTasks::ReadyState chunkState = iter->second->ready(useFlexibleLock);
+    while (chunkState != ChunkTasks::ReadyState::READY
+           && chunkState != ChunkTasks::ReadyState::NO_RESOURCES) {
+        ++iter;
+        if (iter == _chunkMap.end()) {
+            iter = _chunkMap.begin();
+        }
+        if (iter == _activeChunk) {
+            return false;
+        }
+        chunkState = iter->second->ready(useFlexibleLock);
+    }
+    if (chunkState == ChunkTasks::ReadyState::NO_RESOURCES) {
+        // Advancing past a chunk where there aren't enough resources could cause many
+        // scheduling issues.
+        return false;
+    }
+    _readyChunk = iter->second;
+    return true;
+}
+
+
+wbase::Task::Ptr ChunkTaskList::getTask(bool useFlexibleLock) {
+    std::lock_guard<std::mutex> lock(_mapMx);
+    if (_readyChunk == nullptr) {
+        // Attempt to set _readyChunk.
+        _ready(useFlexibleLock);
+    }
+    // If a Task was ready, _readyChunk will not be nullptr.
+    if (_readyChunk != nullptr) {
+        wbase::Task::Ptr task = _readyChunk->getTask(useFlexibleLock);
+        _readyChunk = nullptr;
+        --_taskCount;
+        return task;
+    }
+    return nullptr;
+}
+
+
+/// @return true if _activeChunk will point to a different chunk when getTask is called.
+// This function is normally used by other classes to determine if it is a reasonable time
+// to change priority.
+bool ChunkTaskList::nextTaskDifferentChunkId() {
+    if (_activeChunk == _chunkMap.end()) {
+        return true;
+    }
+    return _activeChunk->second->readyToAdvance();
+}
+
+
+void ChunkTaskList::taskComplete(wbase::Task::Ptr const& task) {
+    auto iter = _chunkMap.find(task->getChunkId());
+    if (iter != _chunkMap.end()) {
+        iter->second->taskComplete(task);
+    }
+}
+
+
+bool ChunkTaskList::setResourceStarved(bool starved) {
+    bool ret = _resourceStarved;
+    _resourceStarved = starved;
+    return ret;
+}
+
+
+void ChunkTasks::SlowTableHeap::push(wbase::Task::Ptr const& task) {
+    _tasks.push_back(task);
+    std::push_heap(_tasks.begin(), _tasks.end(), compareFunc);
+}
+
+
+wbase::Task::Ptr ChunkTasks::SlowTableHeap::pop() {
+    if (_tasks.empty()) {
+        return nullptr;
+    }
+    auto task = _tasks.front();
+    std::pop_heap(_tasks.begin(), _tasks.end(), compareFunc);
+    _tasks.pop_back();
+    return task;
+}
+
+
+/// Queue new Tasks to be run, ordered with the slowest tables first.
```

Comment:
The tables are rated by hand when the when the system is setup. Only shared scan tables must have a rating. The czar passes the rating of all shared scan tables in the query to the worker.